### PR TITLE
Fix missing cases of corruption retries

### PR DESCRIPTION
**Pull request by anand1976**
This PR fixes a few cases where RocksDB was not retrying checksum failure/corruption of file reads with the `verify_and_reconstruct_read` IO option. After fixing these cases, we can almost always successfully open the DB and execute reads even if we see transient corruptions, provided the `FileSystem` supports the `verify_and_reconstruct_read` option. The specific cases fixed in this PR are -
1. CURRENT file
2. IDENTITY file
3. OPTIONS file
4. SST footer

Test plan:
Unit test in `db_io_failure_test.cc` that injects corruption at various stages of DB open and reads

### CODE REVIEW

File: db/db_io_failure_test.cc [hunk 1]
```diff
@@ -970,6 +1072,57 @@ TEST_P(DBIOCorruptionTest, TablePropertiesCorruptionRetry) {
...
+  }
+  Close();
+
+  // DB open will create table readers unless we reduce the table cache
+  // capacity.
+  // SanitizeOptions will set max_open_files to minimum of 20. Table cache
+  // is allocated with max_open_files - 10 as capacity. So override
+  // max_open_files to 11 so table cache capacity will become 1. This will
+  // prevent file open during DB open and force the file to be opened
+  // during MultiGet
```
Comment (jaykorean) [inline db/db_io_failure_test.cc:1098]: This is a nice trick!

File: options/options_parser.cc [hunk 2]
```diff
@@ -268,70 +269,89 @@ Status RocksDBOptionsParser::Parse(const ConfigOptions& config_options_in,
...
+    if (s.ok()) {
+      s = lf_reader.GetStatus();
+    }
+    if (s.ok()) {
+      s = EndSection(config_options, section, title, argument, opt_map);
+      opt_map.clear();
+    }
+    if (s.ok()) {
+      s = ValidityCheck();
+    }
```
Comment (jaykorean) [inline options/options_parser.cc:339]: nit: `if (s.ok()) { break; }` after this validity check would be more readable to me then setting retry = false at the else in line 349. Or you could explicitly exit the loop on line 349.

File: options/options_parser.cc [hunk 3]
```diff
@@ -268,70 +269,89 @@ Status RocksDBOptionsParser::Parse(const ConfigOptions& config_options_in,
...
+    if (!s.ok()) {
+      if ((s.IsCorruption() || s.IsInvalidArgument()) && !retry &&
```
Comment (jaykorean) [inline options/options_parser.cc:341]: Question for my own learning. In what scenario that we'd get `s.IsInvalidArgument()` here and retry with kVerifyAndReconstructRead would succeed?

File: db/db_io_failure_test.cc [hunk 4]
```diff
@@ -101,14 +111,77 @@ class CorruptionFS : public FileSystemWrapper {
...
+          return s;
+        }
+
+        if (opts.verify_and_reconstruct_read) {
+          fs_.MaybeResetOverlapWithCorruptedChunk(fname_, offset_,
+                                                  result->size());
+          return s;
+        }
+
+        // This means the next read after injecting corruption was not
```
Comment (jaykorean) [inline db/db_io_failure_test.cc:160]: wondering if the comment was cut

File: db/db_io_failure_test.cc [hunk 5]
```diff
@@ -970,6 +1072,57 @@ TEST_P(DBIOCorruptionTest, TablePropertiesCorruptionRetry) {
...
+  CorruptionFS* fs =
+      static_cast<CorruptionFS*>(env_guard_->GetFileSystem().get());
+
+  for (int sst = 0; sst < 3; ++sst) {
+    for (int key = 0; key < 100; ++key) {
+      std::stringstream ss;
+      ss << std::setw(3) << 100 * sst + key;
+      ASSERT_OK(Put("key" + ss.str(), "val" + ss.str()));
+    }
+    Flush();
```
Comment (jaykorean) [inline db/db_io_failure_test.cc:1088]: `ASSERT_OK(Flush());`

File: db/db_io_failure_test.cc [hunk 6]
```diff
@@ -970,6 +1072,57 @@ TEST_P(DBIOCorruptionTest, TablePropertiesCorruptionRetry) {
...
+  }
+  Close();
```
Comment (jaykorean) [inline db/db_io_failure_test.cc:1090]: `ASSERT_OK(Close());`

File: options/options_parser.cc (hunk 3 above)
Comment (anand1976) [reply options/options_parser.cc:341]: Any syntax errors during parsing are being considered as `Status::InvalidArgument()`. The syntax error could be due to corruption, and can potentially be corrected by retrying.

File: db/db_io_failure_test.cc (hunk 6 above)
Comment (anand1976) [reply db/db_io_failure_test.cc:1090]: Looks like `DBTestBase::Close()` does not return a Status.

**Review by jaykorean — APPROVED** [review]
👍

**Pull request closed without merging**